concepts/IPLD (#82)
ipfs/ipfs-docs: docs/concepts/ipld.md

# The InterPlanetary Linked Data (IPLD) model

> The Interplanetary Linked Data (IPLD) model is a single namespace for all hash-inspired protocols. This allows for interoperable data formats. For example, you can reference a Git commit to a Bitcoin transaction.

Reviewer: While true, I don't think that should be the first paragraph. The idea of linking between different hash-linked system is again mentioned in the third paragraph. So i would remove the first paragraph.

As mentioned in the [specifications](https://github.com/ipld/specs), the goal of IPLD is to enable decentralized data-structures that are universally addressable and linkable, which in turn will enable more decentralized applications. These data structures allow us to do for data what URLs and links did for HTML web pages.

IPFS, Bitcoin, and other applications use hash-linked data structures ([Merkle trees](https://docs-beta.ipfs.io/concepts/merkle-dag/)) to verify data integrity. However, each protocol has a very different format. IPLD's goal is to provide a data layer definition that allows the linking of different content-addressed data types.

## Where does IPLD fit in?

> IPLD acts as a layer between networking protocols and higher-level applications. This enables developers to build decentralized applications without having to worry about lower-level changes. Both upper- and lower-layer technologies can grow independently, while IPLD acts as a middle layer that resolves the two. This is known as "thin waist" or "narrow waist" protocol.

Reviewer 2: Is there a way to rephase this a little so it doesn't necessarily have to be "between" networking and higher level applications?  It's also possible to use IPLD entirely without a network in the picture.

Kind of similar to git: would you say git is "between" networking and the application of version control?  Not really, right?  Git means you don't have to *care* about networking at all when defining any of the version control semantics -- `git commit` is an entirely local operation; networking doesn't really show up anywhere outside of `git fetch`.  IPLD is similar: making an IPLD object and getting a CID for it is an entirely local operation.

I'm still fine with this paragraph overall.  It's true that we still refer to it as the "thin waist" in a lot of ways.  Just thinking outloud about other more inclusive ways we could try to phrase it in the future.

Author: Good point. I'd be open to any suggestions

![A narrow-waist diagram showing where IPLD fits into the internet ecosystem.](https://gateway.ipfs.io/ipfs/QmXgrfpCcSFfXnXqSz6G3V9E21pTZdBmVdsBQCkr86kHXP)

## The IPLD layer model

The IPLD specifications are divided into [layers](https://github.com/ipld/specs#ipld-layer-model):

0. **Block layer:** The block layer contains the content identifier (CID), which is used to self-describe the multi-base-prefix, version, content type, and content address. This takes the guessing out of encoding/decoding.
1. **Data model layer:** This layer describes what is representable in IPLD (lists, booleans, maps, etc). The kinds of data available are listed in the [`ipld/specs` repository](https://github.com/ipld/specs/blob/master/data-model-layer/data-model.md#kinds).

> 1. **Schema layer:** The schema layer uses the data model layer to build more complex layouts.

Reviewer: It is not clear (if you are not deep into IPLD) what "complex layouts" is supposed to mean. Also the Schema layer is quite a bit more. For more information about it see https://specs.ipld.io/schemas/.

Reviewer 2: :+1: to adding that link.

"complex layouts" is pretty vague, but for a high level overview I think maybe it's fine; I don't have other strong suggestions for how to phrase it, anyway.

Below is the simplified hierarchy:

```

> ┌──────────────────────────────────────────────────┐

Reviewer: Nitpick: not all lines are properly centered.

│                                                  │
│                   2. Schema layer                │
│ (advanced types for multi-block data structures) │
│                                                  │
├──────────────────────────────────────────────────┤
│                                                  │
│                 1. Data model layer              │
│  (basic types for single-block data structures)  │
│                                                  │
├──────────────────────────────────────────────────┤
│                                                  │
│                   0. Block layer                 │
│               (CID, data, codec)                 │
│                                                  │
└──────────────────────────────────────────────────┘
```

## IPLD and "The Merkle Forest"

> The heart of IPLD is to unify data models. Links can be traversed across data formats, opening endless possibilities. Imagine having an Ethereum contract referencing data on IPFS, or having a torrent file link to Zcash. Instead of having one large Merkle tree, as with Bitcoin and Ethereum, IPFS uses IPLD to unite the many Merkle trees into what is coined as the Merkle Forest. 🌳🌲🌴

Reviewer: To me it is not clear how Bitcoin and Ethereum are one large Merkle tree and IPFS is not.

## Further resources

For some examples, you may want to check out the interactive ProtoSchool tutorial [P2P Data Links with Content Addressing](https://proto.school/#/basics/) or learn how to [resolve through IPLD graphs with the dag API](https://github.com/ipfs/js-ipfs/tree/master/examples/traverse-ipld-graphs).

> For more information, you may want to check out the [IPLD explorer](https://explore.ipld.io/) and Juan Benet's [Enter the Merkle Forest](https://www.youtube.com/watch?v=Bqs_LzBjQyk&t=2s).

Reviewer: Should we also link Eric's talk https://media.ccc.de/v/gpn19-105-foundations-for-decentralization-data-with-ipld.

# The InterPlanetary Linked Data (IPLD) model

> The Interplanetary Linked Data (IPLD) model is a set of [specifications](https://github.com/ipld/specs) that unifies all data models that link data with hashes. This allows for interoperable protocols. For example, you can reference a github commit to a bitcoin transaction.

Reviewer 3: ```suggestion
The Interplanetary Linked Data (IPLD) model is a set of [specifications](https://github.com/ipld/specs) that unifies all data models that link data with hashes. This allows for interoperable protocols. For example, you can reference a GitHub commit to a Bitcoin transaction.
```

> IPFS, bitcoin and other applications use hash-linked data structures (Merkle trees) to verify data integrity. However, each protocol has a slightly different format. IPLD's goal is to provide a hash-chain format for distributed data structures.

Reviewer 3: ```suggestion
IPFS, Bitcoin and other applications use hash-linked data structures ([Merkle trees](https://docs-beta.ipfs.io/concepts/merkle-dag/)) to verify data integrity. However, each protocol has a slightly different format. IPLD's goal is to provide a hash-chain format for distributed data structures.
```

> IPLD concerns itself with the data layer of the distributed web. It beings with how data is encoded and decoded. IPLD's [Data Model](https://github.com/ipld/specs) includes standard types such as booleans, integers, lists etc. which are easily representable by common programming languages.

Reviewer 3: ```suggestion
IPLD concerns itself with the data layer of the distributed web. It beings with how data is encoded and decoded. IPLD's [Data Model](https://github.com/ipld/specs) includes standard types such as booleans, integers, lists, etc., which are easily representable by common programming languages.
```

> ## Where Does IPLD Fit In?

Reviewer 3: ```suggestion
## Where does IPLD fit in?
```

> IPLD acts as a layer between networking protocols and higher level applications. This enables developers to build decentralized applications without having to worry about lower level changes.

Reviewer 3: ```suggestion
IPLD acts as a layer between networking protocols and higher-level applications. This enables developers to build decentralized applications without having to worry about lower-level changes.
```

![](https://gateway.ipfs.io/ipfs/QmXgrfpCcSFfXnXqSz6G3V9E21pTZdBmVdsBQCkr86kHXP)

> ## IPLD Layer Model

Reviewer 3: ```suggestion
## The IPLD layer model
```

The IPLD specifications are divided into [layers](https://github.com/ipld/specs#ipld-layer-model):

> 0. Block Layer: The block layer contains the CID which is used to self describe the multibase-prefix, version, content type and content address. This takes the guessing out of encoding/decoding.

Reviewer 3: ```suggestion
0. **Block layer:** The block layer contains the content identifier (CID), which is used to self-describe the multibase-prefix, version, content type and content address. This takes the guessing out of encoding/decoding.
```

> 1. Data Model Layer: This layer describes what is representable in IPLD (lists, booleans, maps etc). Here are all of the [kinds](https://github.com/ipld/specs/blob/master/data-model-layer/data-model.md#kinds).

Reviewer 3: ```suggestion
1. **Data model layer:** This layer describes what is representable in IPLD (lists, booleans, maps, etc). Here are all of the [kinds](https://github.com/ipld/specs/blob/master/data-model-layer/data-model.md#kinds).
```

> 1. Schema Layer: The schema layer uses the data model layer to build more complex layouts.

Reviewer 3: ```suggestion
2. **Schema layer:** The schema layer uses the data model layer to build more complex layouts.
```

Below is the simplified hierarchy:

```
┌──────────────────────────────────────────────────┐
│                                                  │

> │                   2. Schema Layer                │

Reviewer 3: ```suggestion
│                   2. Schema layer                │
```

│ (advanced types for multi-block data structures) │
│                                                  │
├──────────────────────────────────────────────────┤
│                                                  │

> │                 1. Data Model Layer              │

Reviewer 3: ```suggestion
│                 1. Data model layer              │
```

│  (basic types for single-block data structures)  │
│                                                  │
├──────────────────────────────────────────────────┤
│                                                  │

> │                   0. Block Layer                 │

Reviewer 3: ```suggestion
│                   0. Block layer                 │
```

> │               (cid, data, codec)                 │

Reviewer 3: ```suggestion
│               (CID, data, codec)                 │
```

│                                                  │
└──────────────────────────────────────────────────┘
```

> ## Path Traversal Example

Reviewer 3: ```suggestion
## Example: Path traversal
```

> This example will create a simple link from a node to a git commit. The example will use the cli, but this can be done in many ways ([javascript](https://github.com/ipfs/js-ipfs-http-client)).

Reviewer 3: ```suggestion
This example will create a simple link from a node to a Git commit. The example will use the CLI, but this can be done in many other ways (for example, using [JavaScript](https://github.com/ipfs/js-ipfs-http-client)).
```

> 1. Pick a git CID. This example will use [explorer.ipld.io](https://explore.ipld.io/#/explore/z8mWaJHXieAVxxLagBpdaNWFEBKVWmMiE). The goal is to create a node that links to this CID:

Reviewer 3: ```suggestion
1. Pick a Git CID. This example will use [explorer.ipld.io](https://explore.ipld.io/#/explore/z8mWaJHXieAVxxLagBpdaNWFEBKVWmMiE). The goal is to create a node that links to this CID:
```

```
z8mWaJHXieAVxxLagBpdaNWFEBKVWmMiE
```

> 2. Start the ipfs daemon

Reviewer 3: ```suggestion
2. Start the IPFS daemon.
```

```bash
$ ipfs daemon
```

> 2. Create some data. Notice how this data stores the git CID.

Reviewer 3: ```suggestion
2. Create some data. Notice how this data stores the Git CID.
```

```bash
$ echo '{ "message": "linking to git commit", "files": [{ "/": "z8mWaJHXieAVxxLagBpdaNWFEBKVWmMiE" }] }' > test.json
$ cat test.json
{ "message": "linking to git commit", "files": [{ "/": "z8mWaJHXieAVxxLagBpdaNWFEBKVWmMiE" }] }}
```

3. Store an IPLD format node using the [put](https://docs.ipfs.io/reference/api/cli/#ipfs-dag-put) command.

```bash
$ ipfs dag put test.json
bafyreifsnd635mi5ratse2mzg3wget4afcwf2tsop4lymbxtj27tfgizia
```

> 4. From here, use the CID provided to get an IPLD format node.

Reviewer 3: ```suggestion
4. From here, use the CID provided to get an IPLD-format node.
```

```bash
$ ipfs dag get bafyreifsnd635mi5ratse2mzg3wget4afcwf2tsop4lymbxtj27tfgizia
{"files":[{"/":"baf4bcfg4ep767tjp5lxyanx5urpjjgx5q2volvy"}],"message":"linking to git commit"}
```

> 5. Finally, traverse through to access the git commit.

Reviewer 3: ```suggestion
5. Finally, traverse through to access the Git commit.
```

```bash
$ ipfs dag get bafyreifsnd635mi5ratse2mzg3wget4afcwf2tsop4lymbxtj27tfgizia/files/0/
{"author":{"date":"1524708555 +0200","email":"magik6k@users.noreply.github.com","name":"Łukasz Magiera"},"committer":{"date":"1524708555 +0200","email":"noreply@github.com","name":"GitHub"},"message":"Merge pull request #31 from ipfs-shipyard/feat/zip-button\n\nDownload Zip button","other":[" "],"parents":[{"/":"baf4bcfcjvhadrgzmwjflz25kyyc3tolmqaobvfy"},{"/":"baf4bcfcfa2xemiceossebsosuxsi77m5ztatwci"}],"signature":{"Text":" \n wsBcBAABCAAQBQJa4TTLCRBK7hj4Ov3rIwAAdHIIAFIOYsbNvoXO+N7d1pnhR6Ue\n qE2qLrldXeGI+04k3i0rmL8/YAi3zavCA1F1VfjSxX8nYsdYWmjljURiDL07y+5K\n RDxsdHPvf8GUN04rO3mVbCISAme3okGMdrn+hfEdZVdwGdUFva0vtegMWDvnADnY\n si6kURUTONywkNLLHsUD6kPUk2YEZuruaSWblGQwS1ejGzfPo7n4z/ScotC/XxFP\n SCj93XLLlh6HBhh4huJgr2HA/13Tg4XLW5+DnE1mIs3FA0cCQ2WGZCbXEugX/+xv\n u0UWyVv7o8xWt0SJeAFpOLDy7MhCpETJsQ0sFe6Mte6plzsa4vM8/Rq+nI5TzoU=\n =RYjo\n"},"tree":{"/":"baf4bcfcqex3wt5ajfg3lynpufxpiivqt5nxlmxy"}}"
```

Congratulations! We successfuly linked a node with CID `bafyreifsnd...` to `z8mWaJHXieA...`.

> ## IPLD and "The Merkle Forrest"

Reviewer 3: ```suggestion
## IPLD and "The Merkle Forest"
```

Author: Is forest 1 R or 2?

Reviewer 3: One 😊

> The heart of IPLD is to unify data models. Links can be traversed across protocols, opening endless possibilities. Imagine having an Ethereum contract referencing data on IPFS. Or having a torrent file link to Zcash. Instead of having one large Merkle tree like bitcoin and Ethereum, IPFS uses IPLD to unite the many Merkle trees to what is coined as the Merkle Forrest. 🌳🌲🌴

Reviewer 3: ```suggestion
The heart of IPLD is to unify data models. Links can be traversed across protocols, opening endless possibilities. Imagine having an Ethereum contract referencing data on IPFS. Or having a torrent file link to Zcash. Instead of having one large Merkle tree, as with Bitcoin and Ethereum, IPFS uses IPLD to unite the many Merkle trees into what is coined as the Merkle Forest. 🌳🌲🌴
```

# The InterPlanetary Linked Data (IPLD) model

> The Interplanetary Linked Data (IPLD) model is a set of [specifications](https://github.com/ipld/specs) that unifies all data models that link data with hashes. This allows for interoperable protocols. For example, you can reference a GitHub commit to a Bitcoin transaction.

Author: Good point. I made the fixes 👍

> IPFS, Bitcoin and other applications use hash-linked data structures ([Merkle trees](https://docs-beta.ipfs.io/concepts/merkle-dag/)) to verify data integrity. However, each protocol has a slightly different format. IPLD's goal is to provide a hash-chain format for distributed data structures.

Author: fixed

IPLD concerns itself with the data layer of the distributed web. It beings with how data is encoded and decoded. IPLD's [Data Model](https://github.com/ipld/specs) includes standard types such as booleans, integers, lists, etc., which are easily representable by common programming languages.

## Where does IPLD fit in?

IPLD acts as a layer between networking protocols and higher-level applications. This enables developers to build decentralized applications without having to worry about lower-level changes.

> ![](https://gateway.ipfs.io/ipfs/QmXgrfpCcSFfXnXqSz6G3V9E21pTZdBmVdsBQCkr86kHXP)

Author: good idea. I'll think of something this week. If I can't come up with something, then I'll reach out to the team.

{"files":[{"/":"baf4bcfg4ep767tjp5lxyanx5urpjjgx5q2volvy"}],"message":"linking to git commit"}
```

5. Finally, traverse through to access the Git commit.

```bash
$ ipfs dag get bafyreifsnd635mi5ratse2mzg3wget4afcwf2tsop4lymbxtj27tfgizia/files/0/
{"author":{"date":"1524708555 +0200","email":"magik6k@users.noreply.github.com","name":"Łukasz Magiera"},"committer":{"date":"1524708555 +0200","email":"noreply@github.com","name":"GitHub"},"message":"Merge pull request #31 from ipfs-shipyard/feat/zip-button\n\nDownload Zip button","other":[" "],"parents":[{"/":"baf4bcfcjvhadrgzmwjflz25kyyc3tolmqaobvfy"},{"/":"baf4bcfcfa2xemiceossebsosuxsi77m5ztatwci"}],"signature":{"Text":" \n wsBcBAABCAAQBQJa4TTLCRBK7hj4Ov3rIwAAdHIIAFIOYsbNvoXO+N7d1pnhR6Ue\n qE2qLrldXeGI+04k3i0rmL8/YAi3zavCA1F1VfjSxX8nYsdYWmjljURiDL07y+5K\n RDxsdHPvf8GUN04rO3mVbCISAme3okGMdrn+hfEdZVdwGdUFva0vtegMWDvnADnY\n si6kURUTONywkNLLHsUD6kPUk2YEZuruaSWblGQwS1ejGzfPo7n4z/ScotC/XxFP\n SCj93XLLlh6HBhh4huJgr2HA/13Tg4XLW5+DnE1mIs3FA0cCQ2WGZCbXEugX/+xv\n u0UWyVv7o8xWt0SJeAFpOLDy7MhCpETJsQ0sFe6Mte6plzsa4vM8/Rq+nI5TzoU=\n =RYjo\n"},"tree":{"/":"baf4bcfcqex3wt5ajfg3lynpufxpiivqt5nxlmxy"}}"
```

> Congratulations! We successfuly linked a node with CID `bafyreifsnd...` to `z8mWaJHXieA...`.

Author: Can proto.school look at the example? I'd like it to live somewhere.

https://github.com/ipfs/docs/issues/86 wanted some video/code example. Going through this helped me put the pieces together. Especially the key value of "/". At first I was using a different key value "foobar" so my traverse was:

{ "message": "linking to git commit", "files": [{ "foobar": "z8mWaJHXieAVxxLagBpdaNWFEBKVWmMiE" }] }}

<my-cid>/files/0/foobar
returns git-cid
use git-cid to continue traverse
<git-cid>/author...
Not very elegant.

Changing to "/" made the traverse much nicer. 
{ "message": "linking to git commit", "files": [{ "/": "z8mWaJHXieAVxxLagBpdaNWFEBKVWmMiE" }] }}
<my-cid>/files/0/author

> Congratulations! We successfuly linked a node with CID `bafyreifsnd...` to `z8mWaJHXieA...`.

Reviewer 3: @leon-do @hsanjuan -- You both make good points. To that end, what about the following approach?

1. Please explain in text @hsanjuan's point about "how we actually know that something is a Git CID", so that's conceptually covered in the main body text of the doc.

2. Keep the code example, but don't make doing the code example essential to one's understanding of the topic (acid test: what if you were just reading this on your phone on the bus, and couldn't play with code?). One way to do this is to annotate/comment or contextualize *why* the code example is helpful (which you do in your last comment above about putting the pieces together). Another is to relocate it within the content itself so it's closer to the end and a bit more like an appendix: if the reader's eyes glaze over upon seeing a code block and they don't go any further, they should still be able to have enough information to understand the main context.

In the long term, agree that a ProtoSchool tutorial would be awesome -- but developing these is a little more involved, and it would be sad to lose momentum or defer presenting useful information in the meantime. (FYI: Guidelines on building ProtoSchool tutorials are here: https://github.com/ProtoSchool/protoschool.github.io )

Thanks both!

> Congratulations! We successfuly linked a node with CID `bafyreifsnd...` to `z8mWaJHXieA...`.

Author: Sounds good to me @jessicaschilling 

Thoughts @hsanjuan ?

FYI: I'll be happy to write some code for protoschool. I'll have to familiarize myself with the repo first.

> Congratulations! We successfuly linked a node with CID `bafyreifsnd...` to `z8mWaJHXieA...`.

Reviewer 4: My main issue with code examples is that they break silently. The last thing we'll remember as developers is to come fix an example embedded in some documentation page, particularly if we have many. But this is not a problem with the example itself, but with how bad we are at doc maintenance. There are other ways to fix it (a big button "Does it not work? Report a problem with this example" perhaps). If @jessicaschilling wants the example, then that's good, let's keep it.

> Congratulations! We successfuly linked a node with CID `bafyreifsnd...` to `z8mWaJHXieA...`.

Reviewer 3: Agreed about the code example, but also agreed on the opportunity to give feedback. We have the following on the bottom of every page and it's already caught a few problems:
![image](https://user-images.githubusercontent.com/1507828/73681537-c77cfb80-467b-11ea-847d-35d4ded92162.png)
So (for now at least) I'd recommend keeping the code example, downplayed as above. Thanks!

> Congratulations! We successfuly linked a node with CID `bafyreifsnd...` to `z8mWaJHXieA...`.

Reviewer 3: @leon-do PS - @terichadbourne is the ProtoSchool lead and would love to talk to you if you're interested in contributing lesson material!

> Congratulations! We successfuly linked a node with CID `bafyreifsnd...` to `z8mWaJHXieA...`.

Reviewer 5: At first glance it looks to me like these commands to use the DAG API within IPFS are already covered in ProtoSchool. Please see https://proto.school/#/basics (introducing the basic methods) and https://proto.school/#/blog (more complex example of linking data) and let me know if you disagree. If we all agree these are covering the same concepts, let's be sure to link to those tutorials from the docs for further learning. Thanks!

> Congratulations! We successfuly linked a node with CID `bafyreifsnd...` to `z8mWaJHXieA...`.

Author: Hello Teri,

Read https://proto.school/#/basics/03
That covers the example. I'll link to it instead!

# The InterPlanetary Linked Data (IPLD) model

> 

Reviewer 6: For clarification, here is the [contribution rule](https://docs-beta.ipfs.io/community/contribute/grammar-formatting-and-style/#acronyms) explaining @jessicaschilling's point

## Where does IPLD fit in?

> IPLD acts as a layer between networking protocols and higher-level applications. This enables developers to build decentralized applications without having to worry about lower-level changes. Both upper and lower layer technologies can grow independently while IPLD acts as middle layer that resolves the two. This is known as "thin/narrow waist" protocol.

Reviewer 3: ```suggestion
IPLD acts as a layer between networking protocols and higher-level applications. This enables developers to build decentralized applications without having to worry about lower-level changes. Both upper- and lower-layer technologies can grow independently, while IPLD acts as a middle layer that resolves the two. This is known as "thin waist" or "narrow waist" protocol.
```

Reviewer 3: Copy editing 😊

└──────────────────────────────────────────────────┘
```

## IPLD and "The Merkle Forest"

The heart of IPLD is to unify data models. Links can be traversed across data formats, opening endless possibilities. Imagine having an Ethereum contract referencing data on IPFS. Or having a torrent file link to Zcash. Instead of having one large Merkle tree, as with Bitcoin and Ethereum, IPFS uses IPLD to unite the many Merkle trees into what is coined as the Merkle Forest. 🌳🌲🌴

## Further resources

> For some examples, you many want to check out an interactive tutorial [P2P Data Links with Content Addressing](https://proto.school/#/basics/) or learn how to [resolve through IPLD graphs with the dag API](https://github.com/ipfs/js-ipfs/tree/master/examples/traverse-ipld-graphs).

Reviewer 5: ```suggestion
For some examples, you many want to check out the interactive ProtoSchool tutorial [P2P Data Links with Content Addressing](https://proto.school/#/basics/) or learn how to [resolve through IPLD graphs with the dag API](https://github.com/ipfs/js-ipfs/tree/master/examples/traverse-ipld-graphs).
```
